### PR TITLE
Turn off md5sum optionally if content-md5 is not set

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Turn off md5sum optionally if content-md5 is not set

## Motivation and Context
This PR also brings --compat option to run MinIO in strict
S3 compatibility mode, MinIO by default will now try to run
high performance mode.

## Regression
No

## How Has This Been Tested?
Tested with applications which don't require md5sum

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :x: [more...](#4d1e9475-586b-4efa-a755-d3659aa52d31) |
| mint-xl.sh | :x: [more...](#dfe6549c-9613-48b6-adad-d521f725ee96) |
| mint-large-bucket.sh | :x: [more...](#387a1eaf-3a13-4ecf-a3ad-2c92a9489758) |
| mint-compression-dist-xl.sh | :x: [more...](#c880482f-e81a-45d3-98da-9003848ba900) |
| mint-compression-fs.sh | :x: [more...](#a153a079-e5fd-4cb1-8647-d92859db5f11) |
| mint-dist-xl.sh | :x: [more...](#09d95f01-8ce6-4047-abb3-8f58f52c9a64) |
<a name="387a1eaf-3a13-4ecf-a3ad-2c92a9489758"></a>
### 7609-9df7151/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.57:31363
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp cae0cd8f2811:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 2 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 51 seconds
(4/14) Running aws-sdk-ruby tests ... done in 13 seconds
(5/14) Running awscli tests ... done in 2 minutes and 1 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 31 seconds
(8/14) Running minio-dotnet tests ... done in 1 minutes and 11 seconds
(9/14) Running minio-go tests ... done in 2 minutes and 9 seconds
(10/14) Running minio-java tests ... FAILED in 3 minutes and 18 seconds
{
  "name": "minio-java",
  "function": "presignedGetObject(String bucketName, String objectName)",
  "duration": 20182,
  "status": "FAIL",
  "error": "java.net.SocketTimeoutException: timeout >>> [okio.Okio$4.newTimeoutException(Okio.java:232), okio.AsyncTimeout.exit(AsyncTimeout.java:286), okio.AsyncTimeout$2.read(AsyncTimeout.java:241), okio.RealBufferedSource.indexOf(RealBufferedSource.java:358), okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.java:230), okhttp3.internal.http1.Http1Codec.readHeaderLine(Http1Codec.java:226), okhttp3.internal.http1.Http1Codec.readResponseHeaders(Http1Codec.java:202), okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:101), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147), okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121), okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:94), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121), okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147), okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:125), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121), okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:264), okhttp3.RealCall.execute(RealCall.java:93), FunctionalTest.readObject(FunctionalTest.java:183), FunctionalTest.presignedGetObject_test1(FunctionalTest.java:1770), FunctionalTest.runTests(FunctionalTest.java:2997), FunctionalTest.main(FunctionalTest.java:3121)]"
}

Executed 9 out of 14 tests successfully.
```
<a name="a153a079-e5fd-4cb1-8647-d92859db5f11"></a>
### 7609-9df7151/mint-compression-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.59:32749
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp df329a92e321:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 0 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... FAILED in 2 seconds
{
  "name": "aws-sdk-php",
  "function": "createMultipartUpload ( array $params = [] )",
  "args": {
    "Bucket": "aws-sdk-php-5ccb77b5a0762",
    "Object": "obj1"
  },
  "duration": "356",
  "status": "FAIL",
  "error": "Error executing \"CompleteMultipartUpload\" on \"http://72.28.97.59:32749/aws-sdk-php-5ccb77b5a0762/obj1?uploadId=7dc4dec4-8e35-4f19-aa4a-3ce8609e10e5\"; AWS HTTP error: Client error: `POST http://72.28.97.59:32749/aws-sdk-php-5ccb77b5a0762/obj1?uploadId=7dc4dec4-8e35-4f19-aa4a-3ce8609e10e5` resulted in a `400 Bad Request` response:\n<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>InvalidPart</Code><Message>One or more of the specified parts could  (truncated...)\n InvalidPart (client): One or more of the specified parts could not be found.  The part may not have been uploaded, or the specified entity tag may not match the part's entity tag. - <?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>InvalidPart</Code><Message>One or more of the specified parts could not be found.  The part may not have been uploaded, or the specified entity tag may not match the part&#39;s entity tag.</Message><Key>obj1</Key><BucketName>aws-sdk-php-5ccb77b5a0762</BucketName><Resource>/aws-sdk-php-5ccb77b5a0762/obj1</Resource><RequestId>159B0035C1D31423</RequestId><HostId>6cf10a70-9686-467f-8e0a-d58232fa437b</HostId></Error>"
}

Executed 2 out of 14 tests successfully.
```
<a name="c880482f-e81a-45d3-98da-9003848ba900"></a>
### 7609-9df7151/mint-compression-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.54:30358
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 863ce6447dbb:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 0 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... FAILED in 5 seconds
{
  "name": "aws-sdk-php",
  "function": "createMultipartUpload ( array $params = [] )",
  "args": {
    "Bucket": "aws-sdk-php-5ccb77b6e11fc",
    "Object": "obj1"
  },
  "duration": "1026",
  "status": "FAIL",
  "error": "Error executing \"CompleteMultipartUpload\" on \"http://72.28.97.54:30358/aws-sdk-php-5ccb77b6e11fc/obj1?uploadId=642697d3-ad20-444c-aa19-d58943a81126\"; AWS HTTP error: Client error: `POST http://72.28.97.54:30358/aws-sdk-php-5ccb77b6e11fc/obj1?uploadId=642697d3-ad20-444c-aa19-d58943a81126` resulted in a `400 Bad Request` response:\n<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>InvalidPart</Code><Message>One or more of the specified parts could  (truncated...)\n InvalidPart (client): One or more of the specified parts could not be found.  The part may not have been uploaded, or the specified entity tag may not match the part's entity tag. - <?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>InvalidPart</Code><Message>One or more of the specified parts could not be found.  The part may not have been uploaded, or the specified entity tag may not match the part&#39;s entity tag.</Message><Key>obj1</Key><BucketName>aws-sdk-php-5ccb77b6e11fc</BucketName><Resource>/aws-sdk-php-5ccb77b6e11fc/obj1</Resource><RequestId>159B0036CFCFE0C9</RequestId><HostId>d6855ce0-7d37-4664-9f75-a0752e6f8781</HostId></Error>"
}

Executed 2 out of 14 tests successfully.
```
<a name="09d95f01-8ce6-4047-abb3-8f58f52c9a64"></a>
### 7609-9df7151/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.53:32284
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 142b338b868a:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 0 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... FAILED in 44 seconds
{
  "name": "aws-sdk-php",
  "function": "getBucketPolicy ( array $params = [] )",
  "args": {
    "Bucket": "aws-sdk-php-5ccb77b6503ad"
  },
  "duration": "40174",
  "status": "FAIL",
  "error": "Expected getObject to fail with 403"
}

Executed 2 out of 14 tests successfully.
```
<a name="dfe6549c-9613-48b6-adad-d521f725ee96"></a>
### 7609-9df7151/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.60:30893
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 007d31f76542:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 0 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 41 seconds
(4/14) Running aws-sdk-ruby tests ... done in 2 seconds
(5/14) Running awscli tests ... done in 46 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 9 seconds
(8/14) Running minio-dotnet tests ... done in 22 seconds
(9/14) Running minio-go tests ... done in 9 seconds
(10/14) Running minio-java tests ... done in 15 seconds
(11/14) Running minio-js tests ... done in 35 seconds
(12/14) Running minio-py tests ... FAILED in 1 minutes and 25 seconds
{
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1226, in test_presigned_get_object_expiry_5sec\n    raise ValueError('Presigned get url failed to expire!')\nValueError: Presigned get url failed to expire!\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1885, in main\n    test_presigned_get_object_expiry_5sec(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1228, in test_presigned_get_object_expiry_5sec\n    raise Exception(err)\nException: Presigned get url failed to expire!\n",
  "function": "presigned_get_object(bucket_name, object_name, expires, response_headers, request_date)",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-a48c3b1a-a430-41e1-9c51-cc64b10f52fa",
    "object_name": "1b624e11-63bc-4e4c-9749-480fda7aeb05"
  },
  "duration": 5024,
  "name": "minio-py:test_presigned_get_object_expiry_5sec",
  "message": "Presigned get url failed to expire!"
}

Executed 11 out of 14 tests successfully.
```
<a name="4d1e9475-586b-4efa-a755-d3659aa52d31"></a>
### 7609-9df7151/mint-compression-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.54:32296
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 8c231871c6e5:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 0 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... FAILED in 2 seconds
{
  "name": "aws-sdk-php",
  "function": "createMultipartUpload ( array $params = [] )",
  "args": {
    "Bucket": "aws-sdk-php-5ccb77b5af783",
    "Object": "obj1"
  },
  "duration": "534",
  "status": "FAIL",
  "error": "Error executing \"CompleteMultipartUpload\" on \"http://72.28.97.54:32296/aws-sdk-php-5ccb77b5af783/obj1?uploadId=9b69fc00-ba17-4e23-ad38-1dcde19c2213\"; AWS HTTP error: Client error: `POST http://72.28.97.54:32296/aws-sdk-php-5ccb77b5af783/obj1?uploadId=9b69fc00-ba17-4e23-ad38-1dcde19c2213` resulted in a `400 Bad Request` response:\n<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>InvalidPart</Code><Message>One or more of the specified parts could  (truncated...)\n InvalidPart (client): One or more of the specified parts could not be found.  The part may not have been uploaded, or the specified entity tag may not match the part's entity tag. - <?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>InvalidPart</Code><Message>One or more of the specified parts could not be found.  The part may not have been uploaded, or the specified entity tag may not match the part&#39;s entity tag.</Message><Key>obj1</Key><BucketName>aws-sdk-php-5ccb77b5af783</BucketName><Resource>/aws-sdk-php-5ccb77b5af783/obj1</Resource><RequestId>159B0035CB751D03</RequestId><HostId>46badc53-692d-4235-b590-66a0ca84554a</HostId></Error>"
}

Executed 2 out of 14 tests successfully.
```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :x: [more...](#a6137aa4-62e6-49d5-af6e-4591b0144528) |
<a name="a6137aa4-62e6-49d5-af6e-4591b0144528"></a>
### 7609-fa80aea/mint-compression-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.51:31076
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 94cb75b0967f:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 6 minutes and 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-e4gg29kpnspgpk",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 360010,
  "error": "XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 159B03F5D00E89D1, host id: ",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |

**Comment by krishnasrinivas** [comment]
@harshavardhana did we decide anything on `--compat` option?

**Comment by poornas** [comment]
this will break S3 API guarantees on md5sum for encrypted objects
```
(base) ➜  ~ aws s3api put-object --bucket hidebucket --key issue --server-side-encryption AES256 --endpoint-url http://localhost:9000
{
    "ETag": "\"20000f0080e4bbd3fe7bfab730d3db914ce47718b408699b8958c9ed4c1fe7197977a819414c8de2e7aaf3fa9712d8ee-1\"",
    "ServerSideEncryption": "AES256"

```

**Comment by harshavardhana** [comment]
> (base) ➜  ~ aws s3api put-object --bucket hidebucket --key issue --server-side-encryption AES256 --endpoint-url http://localhost:9000
> {
>     "ETag": "\"20000f0080e4bbd3fe7bfab730d3db914ce47718b408699b8958c9ed4c1fe7197977a819414c8de2e7aaf3fa9712d8ee-1\"",
>     "ServerSideEncryption": "AES256"

Yes we decided on compat option @krishnasrinivas - it is agreed that we will break compatibility here @poornas and it will be available as `--compat` flag for strict compatibility.

**Comment by harshavardhana** [comment]
And another thing to remember is if you didn't set content-md5 server is not expected to compute md5 etag @poornas

**Comment by harshavardhana** [comment]
PR is updated with `--compat` option which will ensure MinIO runs in strict S3 mode. 

Scenarios when ETag is not guaranteed to be hex version of MD5Sum

- SSE-C PUT requests
- Multipart requests
- All PUT requests without --compat mode
- SSE-S3 requests without --compat mode
- The server has a compression feature on

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |

**Review by poornas — CHANGES_REQUESTED** [review]
Tested , LGTM - can you document this as well

**Comment by harshavardhana** [comment]
Where do you suggest we document this?

**Comment by poornas** [comment]
@y4m4, README - this is a breaking change since default server mode is turning off strict S3 compatibility

File: cmd/object-api-utils.go [hunk 1]
```diff
@@ -675,9 +675,26 @@ func (p *PutObjReader) Size() int64 {
...
+		appendHyphen = true
+	}
 	if p.sealMD5Fn != nil {
 		md5sumCurr = p.sealMD5Fn(md5sumCurr)
 	}
+	if appendHyphen {
+		// Make sure to return etag string upto 32 length, for SSE
+		// requests ETag might be longer and the code decrypting the
+		// ETag ignores ETag in multipart ETag form i.e <hex>-N
+		return hex.EncodeToString(md5sumCurr)[:32] + "-1"
```
Comment (krishnasrinivas) [inline cmd/object-api-utils.go:696]: if sealed md5 is modified it can't be decrypted anymore

Comment (harshavardhana) [reply cmd/object-api-utils.go:696]: it can @krishnasrinivas the stored value is not this value

**Review by poornas — APPROVED** [review]
Tested, LGTM

**Review by krishnasrinivas — APPROVED** [review]

**Comment by flixr** [comment]
Couldn't find this in the documentation, where is this documented now?

**Pull request merged**